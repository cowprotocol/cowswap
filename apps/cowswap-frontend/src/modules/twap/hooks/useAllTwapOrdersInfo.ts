import { useEffect, useMemo, useRef, useState } from 'react'

import { TwapOrderInfo, TwapOrdersSafeData } from '../types'
import { getConditionalOrderId } from '../utils/getConditionalOrderId'
import { parseTwapOrderStruct } from '../utils/parseTwapOrderStruct'

export function useAllTwapOrdersInfo(ordersSafeData: TwapOrdersSafeData[]): TwapOrderInfo[] {
  const [allOrdersInfo, setAllOrdersInfo] = useState<Record<string, TwapOrderInfo>>({})
  const allOrdersInfoRef = useRef(allOrdersInfo)

  // eslint-disable-next-line react-hooks/refs
  allOrdersInfoRef.current = allOrdersInfo

  useEffect(() => {
    const parsed = parseOrdersSafeData(ordersSafeData)
    const ordersToAdd = parsed.filter((item) => !allOrdersInfoRef.current[item.id])

    if (!ordersToAdd.length) return

    const update = ordersToAdd.reduce<Record<string, TwapOrderInfo>>((acc, item) => {
      acc[item.id] = item

      return acc
    }, {})

    setAllOrdersInfo((state) => {
      return { ...state, ...update }
    })
  }, [ordersSafeData])

  return useMemo(() => Object.values(allOrdersInfo), [allOrdersInfo])
}

function parseOrdersSafeData(ordersSafeData: TwapOrdersSafeData[]): TwapOrderInfo[] {
  const ordersInfoMap = ordersSafeData.reduce<{ [id: string]: TwapOrderInfo }>((acc, data) => {
    try {
      const id = getConditionalOrderId(data.conditionalOrderParams)
      const existingOrder = acc[id]

      /**
       * There might be two Safe transactions with the same order inside.
       * But only one of them will be executed.
       *
       * For example, you propose a transaction with TWAP order and execute it.
       * Then, you propose another transaction with the same TWAP order.
       * After you realize that the proposed transaction is a duplicate, and you replace it or cancel.
       * In this case we should skip the second transaction, because the first one is already executed.
       */
      if (existingOrder?.safeData.safeTxParams.isExecuted) {
        return acc
      }

      const info = {
        id,
        orderStruct: parseTwapOrderStruct(data.conditionalOrderParams.staticInput),
        safeData: data,
      }

      acc[id] = info
    } catch {
      // Do nothing
    }

    return acc
  }, {})

  return Object.values(ordersInfoMap)
}
