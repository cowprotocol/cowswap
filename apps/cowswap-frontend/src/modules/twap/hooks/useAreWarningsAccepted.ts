import { useAtomValue } from 'jotai'

import { useIsNoImpactWarningAccepted } from 'modules/trade'

import { useIsFallbackHandlerRequired } from './useFallbackHandlerVerification'

import { twapOrdersSettingsAtom } from '../state/twapOrdersSettingsAtom'

export function useAreWarningsAccepted(): boolean {
  const { isFallbackHandlerSetupAccepted } = useAtomValue(twapOrdersSettingsAtom)
  const isNoImpactWarningAccepted = useIsNoImpactWarningAccepted()
  const isFallbackHandlerRequired = useIsFallbackHandlerRequired()

  if (isFallbackHandlerRequired) return isFallbackHandlerSetupAccepted

  return isNoImpactWarningAccepted
}
