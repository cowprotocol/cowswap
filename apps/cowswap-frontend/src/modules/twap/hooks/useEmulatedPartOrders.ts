import { useAtomValue } from 'jotai'
import { useMemo } from 'react'

import { useMachineTimeMs } from '@cowprotocol/common-hooks'
import { TokensByAddress } from '@cowprotocol/tokens'

import ms from 'ms.macro'

import { Order } from 'legacy/state/orders/actions'

import { twapOrdersAtom, TwapOrdersList } from '../state/twapOrdersListAtom'
import { TwapPartOrderItem, twapPartOrdersListAtom } from '../state/twapPartOrdersAtom'
import { emulatePartAsOrder } from '../utils/emulatePartAsOrder'
import { mapPartOrderToStoreOrder } from '../utils/mapPartOrderToStoreOrder'

const EMULATED_ORDERS_REFRESH_MS = ms`5s`

export function useEmulatedPartOrders(tokensByAddress: TokensByAddress | undefined): Order[] {
  const twapOrders = useAtomValue(twapOrdersAtom)
  const twapParticleOrders = useAtomValue(twapPartOrdersListAtom)
  // Update emulated part orders every 5 seconds to recalculate expired state
  const refresher = useMachineTimeMs(EMULATED_ORDERS_REFRESH_MS)

  return useMemo(() => {
    // It's not possible, just to prevent react-hooks/exhaustive-deps errors
    if (!refresher) return []
    if (!tokensByAddress) return []

    return emulatePartOrders(twapParticleOrders, twapOrders, tokensByAddress)
  }, [twapParticleOrders, twapOrders, tokensByAddress, refresher])
}

function emulatePartOrders(
  twapParticleOrders: TwapPartOrderItem[],
  twapOrders: TwapOrdersList,
  tokensByAddress: TokensByAddress
): Order[] {
  return twapParticleOrders.reduce<Order[]>((acc, item) => {
    if (item.isCreatedInOrderBook) return acc

    const isVirtualPart = true
    const parent = twapOrders[item.twapOrderId]

    if (!parent) return acc

    const enrichedOrder = emulatePartAsOrder(item, parent)
    const order = mapPartOrderToStoreOrder(item, enrichedOrder, isVirtualPart, parent, tokensByAddress)

    if (order) acc.push(order)

    return acc
  }, [])
}
