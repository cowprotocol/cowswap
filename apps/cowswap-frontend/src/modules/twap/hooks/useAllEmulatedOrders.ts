import { useMemo } from 'react'

import { UiOrderType } from '@cowprotocol/types'
import { useIsSafeApp, useWalletInfo } from '@cowprotocol/wallet'

import { Order } from 'legacy/state/orders/actions'
import { useOrders } from 'legacy/state/orders/hooks'

import { useEmulatedPartOrders } from './useEmulatedPartOrders'
import { useEmulatedTwapOrders } from './useEmulatedTwapOrders'
import { useTwapOrdersTokens } from './useTwapOrdersTokens'

export function useAllEmulatedOrders(): Order[] {
  const { chainId, account } = useWalletInfo()
  const twapOrdersTokens = useTwapOrdersTokens()
  const emulatedTwapOrders = useEmulatedTwapOrders(twapOrdersTokens)
  const emulatedPartOrders = useEmulatedPartOrders(twapOrdersTokens)
  const isSafeApp = useIsSafeApp()

  const twapOrders = useOrders(chainId, account, UiOrderType.TWAP)
  const discreteTwapOrders = useMemo(() => {
    return twapOrders.filter((order) => order.composableCowInfo?.isVirtualPart === false)
  }, [twapOrders])

  return useMemo(() => {
    if (!isSafeApp) return []

    return emulatedTwapOrders.concat(emulatedPartOrders).concat(discreteTwapOrders)
  }, [emulatedTwapOrders, emulatedPartOrders, discreteTwapOrders, isSafeApp])
}
