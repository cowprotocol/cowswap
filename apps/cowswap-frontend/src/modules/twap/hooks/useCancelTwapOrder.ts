import { useAtomValue, useSetAtom } from 'jotai'
import { useCallback } from 'react'

import { useSafeAppsSdk } from '@cowprotocol/wallet'

import { Order } from 'legacy/state/orders/actions'

import { useComposableCowContract } from 'modules/advancedOrders/hooks/useComposableCowContract'

import type { OnChainCancellation } from 'common/hooks/useCancelOrder/onChainCancellation'
import { useGP2SettlementContract } from 'common/hooks/useContract'

import { cancelTwapOrderTxs, estimateCancelTwapOrderTxs } from '../services/cancelTwapOrderTxs'
import { setTwapOrderStatusAtom } from '../state/twapOrdersListAtom'
import { twapPartOrdersAtom } from '../state/twapPartOrdersAtom'
import { TwapOrderStatus } from '../types'

export function useCancelTwapOrder(): (twapOrderId: string, order: Order) => Promise<OnChainCancellation> {
  const twapPartOrders = useAtomValue(twapPartOrdersAtom)
  const setTwapOrderStatus = useSetAtom(setTwapOrderStatusAtom)
  const safeAppsSdk = useSafeAppsSdk()
  const settlementContract = useGP2SettlementContract()
  const composableCowContract = useComposableCowContract()

  return useCallback(
    async (twapOrderId: string, order: Order) => {
      if (!composableCowContract || !settlementContract || !safeAppsSdk) {
        throw new Error('Context is not full to cancel TWAP order')
      }

      const partOrder = twapPartOrders[twapOrderId]?.sort((a, b) => a.order.validTo - b.order.validTo)[0]
      const partOrderId = partOrder?.uid

      const context = { composableCowContract, settlementContract, orderId: twapOrderId, partOrderId }

      return {
        estimatedGas: await estimateCancelTwapOrderTxs(context),
        sendTransaction: (processCancelledOrder) => {
          return safeAppsSdk.txs.send({ txs: cancelTwapOrderTxs(context) }).then((res) => {
            const txHash = res.safeTxHash
            const sellTokenAddress = order.inputToken.address
            const sellTokenSymbol = order.inputToken.symbol

            setTwapOrderStatus(twapOrderId, TwapOrderStatus.Cancelling)
            processCancelledOrder({ txHash, orderId: twapOrderId, sellTokenAddress, sellTokenSymbol })
          })
        },
      }
    },
    [composableCowContract, settlementContract, safeAppsSdk, twapPartOrders, setTwapOrderStatus]
  )
}
