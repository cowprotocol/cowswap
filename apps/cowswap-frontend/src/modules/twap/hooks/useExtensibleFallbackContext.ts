import { GPv2Settlement } from '@cowprotocol/abis'
import { SupportedChainId } from '@cowprotocol/cow-sdk'
import { useWalletInfo } from '@cowprotocol/wallet'
import { Web3Provider } from '@ethersproject/providers'
import { useWeb3React } from '@web3-react/core'

import { useGP2SettlementContract } from 'common/hooks/useContract'

export interface ExtensibleFallbackContext {
  safeAddress: string
  chainId: SupportedChainId
  settlementContract: GPv2Settlement
  provider: Web3Provider
}

export function useExtensibleFallbackContext(): ExtensibleFallbackContext | null {
  const { chainId, account } = useWalletInfo()
  const settlementContract = useGP2SettlementContract()
  const { provider } = useWeb3React()

  if (!account || !settlementContract || !provider || !chainId) return null

  return { settlementContract, provider, chainId, safeAddress: account }
}
