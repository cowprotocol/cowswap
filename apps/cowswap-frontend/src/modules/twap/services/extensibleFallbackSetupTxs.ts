import { MetaTransactionData } from '@safe-global/safe-core-sdk-types'

import { COMPOSABLE_COW_ADDRESS, SAFE_EXTENSIBLE_HANDLER_ADDRESS } from 'modules/advancedOrders/const'

import { getSignatureVerifierContract } from './getSignatureVerifierContract'

import { ExtensibleFallbackContext } from '../hooks/useExtensibleFallbackContext'

export async function extensibleFallbackSetupTxs(context: ExtensibleFallbackContext): Promise<MetaTransactionData[]> {
  const { chainId, safeAddress, settlementContract } = context

  const domainSeparator = await settlementContract.callStatic.domainSeparator()

  const signatureVerifierContract = await getSignatureVerifierContract(context)
  const composableCowContractAddress = COMPOSABLE_COW_ADDRESS[chainId]
  const extensibleHandlerAddress = SAFE_EXTENSIBLE_HANDLER_ADDRESS[chainId]

  const setFallbackHandlerTx = {
    to: safeAddress,
    data: signatureVerifierContract.interface.encodeFunctionData('setFallbackHandler', [extensibleHandlerAddress]),
    value: '0',
    operation: 0,
  }

  const setDomainVerifierTx = {
    to: safeAddress,
    data: signatureVerifierContract.interface.encodeFunctionData('setDomainVerifier', [
      domainSeparator,
      composableCowContractAddress,
    ]),
    value: '0',
    operation: 0,
  }

  return [setFallbackHandlerTx, setDomainVerifierTx]
}
