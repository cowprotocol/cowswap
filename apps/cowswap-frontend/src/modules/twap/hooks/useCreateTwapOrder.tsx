import { useAtomValue, useSetAtom } from 'jotai'
import { useCallback } from 'react'

import { orderAnalytics, twapConversionAnalytics } from '@cowprotocol/analytics'
import { OrderKind } from '@cowprotocol/cow-sdk'
import { UiOrderType } from '@cowprotocol/types'
import { useSafeAppsSdk, useWalletInfo } from '@cowprotocol/wallet'
import { CurrencyAmount, Token } from '@uniswap/sdk-core'

import { Nullish } from 'types'

import { updateAdvancedOrdersAtom, useAdvancedOrdersDerivedState } from 'modules/advancedOrders'
import { useAppData, useUploadAppData } from 'modules/appData'
import { emitPostedOrderEvent } from 'modules/orders'
import { getCowSoundSend } from 'modules/sounds'
import { useTradeConfirmActions, useTradePriceImpact } from 'modules/trade'
import { TradeFlowAnalyticsContext, tradeFlowAnalytics } from 'modules/trade/utils/analytics'

import { useConfirmPriceImpactWithoutFee } from 'common/hooks/useConfirmPriceImpactWithoutFee'

import { useExtensibleFallbackContext } from './useExtensibleFallbackContext'
import { useTwapOrderCreationContext } from './useTwapOrderCreationContext'

import { DEFAULT_TWAP_EXECUTION_INFO } from '../const'
import { createTwapOrderTxs } from '../services/createTwapOrderTxs'
import { extensibleFallbackSetupTxs } from '../services/extensibleFallbackSetupTxs'
import { twapOrderAtom } from '../state/twapOrderAtom'
import { addTwapOrderToListAtom } from '../state/twapOrdersListAtom'
import { TwapOrderItem, TwapOrderStatus } from '../types'
import { buildTwapOrderParamsStruct } from '../utils/buildTwapOrderParamsStruct'
import { getConditionalOrderId } from '../utils/getConditionalOrderId'
import { getErrorMessage } from '../utils/parseTwapError'
import { twapOrderToStruct } from '../utils/twapOrderToStruct'

export function useCreateTwapOrder() {
  const { chainId, account } = useWalletInfo()
  const twapOrder = useAtomValue(twapOrderAtom)
  const addTwapOrderToList = useSetAtom(addTwapOrderToListAtom)

  const { inputCurrencyAmount, outputCurrencyAmount } = useAdvancedOrdersDerivedState()

  const appDataInfo = useAppData()
  const safeAppsSdk = useSafeAppsSdk()
  const twapOrderCreationContext = useTwapOrderCreationContext(inputCurrencyAmount as Nullish<CurrencyAmount<Token>>)
  const extensibleFallbackContext = useExtensibleFallbackContext()

  const updateAdvancedOrdersState = useSetAtom(updateAdvancedOrdersAtom)

  const tradeConfirmActions = useTradeConfirmActions()
  const uploadAppData = useUploadAppData()

  const { priceImpact } = useTradePriceImpact()
  const { confirmPriceImpactWithoutFee } = useConfirmPriceImpactWithoutFee()

  return useCallback(
    async (fallbackHandlerIsNotSet: boolean) => {
      if (!chainId || !account) return
      if (
        !inputCurrencyAmount ||
        !outputCurrencyAmount ||
        !twapOrderCreationContext ||
        !extensibleFallbackContext ||
        !safeAppsSdk ||
        !appDataInfo ||
        !twapOrder
      )
        return

      const isPriceImpactConfirmed = await confirmPriceImpactWithoutFee(priceImpact)

      if (!isPriceImpactConfirmed) {
        return
      }

      const pendingTrade = {
        inputAmount: inputCurrencyAmount,
        outputAmount: outputCurrencyAmount,
      }

      const orderType = UiOrderType.TWAP
      const twapFlowAnalyticsContext: TradeFlowAnalyticsContext = {
        account,
        recipient: twapOrder.receiver,
        recipientAddress: twapOrder.receiver,
        marketLabel: [inputCurrencyAmount.currency.symbol, outputCurrencyAmount.currency.symbol].join(','),
        orderType,
      }

      try {
        const paramsStruct = buildTwapOrderParamsStruct(chainId, twapOrder)
        const orderId = getConditionalOrderId(paramsStruct)

        tradeConfirmActions.onSign(pendingTrade)
        tradeFlowAnalytics.placeAdvancedOrder(twapFlowAnalyticsContext)
        twapConversionAnalytics('posted', fallbackHandlerIsNotSet)

        const fallbackSetupTxs = fallbackHandlerIsNotSet
          ? await extensibleFallbackSetupTxs(extensibleFallbackContext)
          : []
        const createOrderTxs = createTwapOrderTxs(twapOrder, paramsStruct, twapOrderCreationContext)
        const { safeTxHash } = await safeAppsSdk.txs.send({ txs: [...fallbackSetupTxs, ...createOrderTxs] })

        const orderItem: TwapOrderItem = {
          order: twapOrderToStruct(twapOrder),
          status: TwapOrderStatus.WaitSigning,
          chainId,
          safeAddress: account,
          submissionDate: new Date().toISOString(),
          id: orderId,
          executionInfo: { confirmedPartsCount: 0, info: DEFAULT_TWAP_EXECUTION_INFO },
        }

        addTwapOrderToList(orderItem)

        getCowSoundSend().play()

        emitPostedOrderEvent({
          chainId,
          id: orderId,
          orderCreationHash: safeTxHash,
          kind: OrderKind.SELL,
          receiver: twapOrder.receiver,
          inputAmount: twapOrder.sellAmount,
          outputAmount: twapOrder.buyAmount,
          owner: account,
          uiOrderType: orderType,
        })

        orderAnalytics('Posted', orderType, 'Presign')

        uploadAppData({ chainId, orderId, appData: appDataInfo })
        updateAdvancedOrdersState({ recipient: null, recipientAddress: null })
        tradeConfirmActions.onSuccess(safeTxHash)
        tradeFlowAnalytics.sign(twapFlowAnalyticsContext)
        twapConversionAnalytics('signed', fallbackHandlerIsNotSet)
      } catch (error) {
        console.error('[useCreateTwapOrder] error', error)
        const errorMessage = getErrorMessage(error)
        tradeConfirmActions.onError(errorMessage)
        tradeFlowAnalytics.error(error, errorMessage, twapFlowAnalyticsContext)
        twapConversionAnalytics('rejected', fallbackHandlerIsNotSet)
      }
    },
    [
      chainId,
      account,
      inputCurrencyAmount,
      outputCurrencyAmount,
      twapOrderCreationContext,
      extensibleFallbackContext,
      safeAppsSdk,
      appDataInfo,
      twapOrder,
      confirmPriceImpactWithoutFee,
      priceImpact,
      tradeConfirmActions,
      addTwapOrderToList,
      uploadAppData,
      updateAdvancedOrdersState,
    ]
  )
}
