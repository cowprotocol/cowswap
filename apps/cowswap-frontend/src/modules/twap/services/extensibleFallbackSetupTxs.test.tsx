import { SupportedChainId } from '@cowprotocol/cow-sdk'
import { Web3Provider } from '@ethersproject/providers'

import { extensibleFallbackSetupTxs } from './extensibleFallbackSetupTxs'

import { ExtensibleFallbackContext } from '../hooks/useExtensibleFallbackContext'

describe('extensibleFallbackSetupTxs - service to generate transactions for ExtensibleFallback setup', () => {
  it('Should create a bundle of two transactions: setFallbackHandler and setDomainVerifier', async () => {
    const context: ExtensibleFallbackContext = {
      chainId: SupportedChainId.SEPOLIA,
      safeAddress: '0xA12D770028d7072b80BAEb6A1df962cccfd1dddd',
      settlementContract: {
        callStatic: { domainSeparator: () => '0xa5b986c2f5845d520bcb903639360b147735589732066cea24a3a59678025c94' },
      } as any,
      provider: new Web3Provider(() => Promise.resolve(null)),
    }

    const result = await extensibleFallbackSetupTxs(context)

    expect(result).toMatchSnapshot()
  })
})
