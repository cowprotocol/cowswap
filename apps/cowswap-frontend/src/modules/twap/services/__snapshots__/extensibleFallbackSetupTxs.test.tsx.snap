// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extensibleFallbackSetupTxs - service to generate transactions for ExtensibleFallback setup Should create a bundle of two transactions: setFallbackHandler and setDomainVerifier 1`] = `
[
  {
    "data": "0xf08a03230000000000000000000000002f55e8b20d0b9fefa187aa7d00b6cbe563605bf5",
    "operation": 0,
    "to": "0xA12D770028d7072b80BAEb6A1df962cccfd1dddd",
    "value": "0",
  },
  {
    "data": "0x3365582ca5b986c2f5845d520bcb903639360b147735589732066cea24a3a59678025c94000000000000000000000000fdafc9d1902f4e0b84f65f49f244b32b31013b74",
    "operation": 0,
    "to": "0xA12D770028d7072b80BAEb6A1df962cccfd1dddd",
    "value": "0",
  },
]
`;
